server/src/modules/users/domain/value-objects/encrypted-secret.ts

```diff
@@ -0,0 +1,64 @@
+import bcrypt from 'bcrypt'
+
+import { Result } from '../../../../shared/core/result'
+import { ValueObject } from '../../../../shared/domain/value-object'
+import { UserValueObjectErrors } from './errors'
+
+export interface EncryptedSecretProps {
+  value: string
+  hashed: boolean
+}
+
+//The protected and public members of this base class can be inherited as needed
+export abstract class EncryptedSecret extends ValueObject<EncryptedSecretProps> {
+  protected static minLength: number = 64
+
+  public constructor(props: EncryptedSecretProps) {
+    super(props)
+  }
+
+  protected static isAppropriateLength(secret: string): boolean {
+    return secret.length >= this.minLength
+  }
+
+  get value(): string {
+    return this.props.value
+  }
+
+  public isAlreadyHashed(): boolean {
+    return this.props.hashed
+  }
+
+  public getHashedValue(): Promise<string> {
+    return new Promise((resolve) => {
+      if (this.isAlreadyHashed()) {
+        return resolve(this.props.value)
+      } else {
+        return resolve(this.hashSecret(this.props.value))
+      }
+    })
+  }
+
+  protected hashSecret(secret: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+      bcrypt.hash(secret, 10, (err, hash) => {
+        if (err) return reject(err)
+        resolve(hash)
+      })
+    })
+  }
```

Comment:
I'm guessing bcrypt's use of callbacks prevents us from using the regular `async` syntax here?

Reply:
Using `await bcrypt.hash...` doesn't allow for easy detection of Promise rejections in the case of a hashing error (both the resolved and rejected values are strings), so yeh.